src/test/java/com/microsoft/aad/msal4j/InstanceDiscoveryTest.java

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import org.powermock.api.easymock.PowerMock;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.testng.PowerMockTestCase;
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+@PrepareForTest(AadInstanceDiscovery.class)
+public class InstanceDiscoveryTest extends PowerMockTestCase {
+
+    @BeforeMethod
+    public void setup(){
+        AadInstanceDiscovery.cache.clear();
+    }
+
+    @Test
+    public void instanceDiscoveryMetadataTest_useDataInTokenCache() throws Exception {
+        // properly formatted instance discovery metadata in token cache
+        String tokenCacheData = TestHelper.readResource(
+                this.getClass(),
+                "/Instance_discovery_metadata_cache_data/cache_instance_discovery_metadata_valid.json");
+
+        PublicClientApplication app = PublicClientApplication.builder("client_id")
+                .correlationId("correlation_id")
+                .authority("https://login.microsoftonline.com/my_tenant")
+                .build();
+
+        app.tokenCache().deserialize(tokenCacheData);
+
+        AuthorizationCodeParameters parameters = AuthorizationCodeParameters.builder(
+                "code", new URI("http://my.redirect.com")).build();
+
+        MsalRequest msalRequest = new AuthorizationCodeRequest(
+                parameters,
+                app,
+                new RequestContext(app, PublicApi.ACQUIRE_TOKEN_BY_AUTHORIZATION_CODE));
+
+        URL authority = new URL(app.authority());
+
+        PowerMock.mockStaticPartial(AadInstanceDiscovery.class, "sendInstanceDiscoveryRequest");
```

Comment:
[Aside] If you use an interface and either make your service a singleton or just pass the instance around, the code would be easier to test without tools for "untestable code"